api/dbadapters/sqldatawarehouse.ts

```diff
@@ -0,0 +1,111 @@
+import { Credentials } from "@dataform/api/commands/credentials";
+import { DbAdapter, OnCancel } from "@dataform/api/dbadapters/index";
+import { dataform } from "@dataform/protos";
+
+const sql = require("mssql");
+
+interface IRequest {
+  query: (query: string) => Promise<{ rows: any[] }>;
+}
+
+interface ISQLDataWarehouseConfig {
+  server?: string;
+  port?: number;
+  user?: string;
+  password?: string;
+  database?: string;
+  encrypt: boolean;
+}
+
+export class SQLDataWarehouseDBAdapter implements DbAdapter {
+  private pool: any;
+
+  constructor(credentials: Credentials) {
+    const sqlDataWarehouseCredentials = credentials as dataform.ISQLDataWarehouse;
+    const config: ISQLDataWarehouseConfig = {
+      server: sqlDataWarehouseCredentials.server,
+      port: sqlDataWarehouseCredentials.port,
+      user: sqlDataWarehouseCredentials.username,
+      password: sqlDataWarehouseCredentials.password,
+      database: sqlDataWarehouseCredentials.databaseName,
+      encrypt: true
+    };
+    const conn = new sql.ConnectionPool(config).connect();
+
+    conn.then(pool =>{
+      pool.on('error',err =>{
+        throw new Error(err);
+      })
+    })
+
+    this.pool = conn;
+  }
+
+  public async execute(statement: string, onCancel?: OnCancel) {
+    let pool = await this.pool;
+    let request = pool.request();
+
+    if(onCancel){
+      onCancel(()=>{
+        request.cancel();
+      })
+    }
+
+    return request
+          .query(statement)
+          .then(result => result.recordset);
+  }
+
+  public async evaluate(statement: string) {
+    let pool = await this.pool;
+    return pool.request()
+          .query(`explain ${statement}`)
+          .then(() => {});
+  }
+
+  public async tables(): Promise<dataform.ITarget[]> {
+    let result = await this.execute(`select table_name, table_schema from information_schema.tables`)
+    return result.map(row => ({
+      schema: row.table_schema,
+      name: row.table_name
+    }))
+  }
+
+  public table(target: dataform.ITarget): Promise<dataform.ITableMetadata> {
+    return Promise.all([
```

Comment:
this can be something like:

```
const [columnRows, tableRows] = await Promise.all([ ... ]);
```

Reply:
this code is exactly the same for redshift and snowflake. i dont want sql to be the odd man out :). if you want to refactor to create a base for sql adapters?